Updated MaskValueSet to match EnumSet better

JaniruTEC opened the pull request:
Added additional convenience methods to MaskValueSet (allOf, copyOf, of)
Added additional methods for copying MaskValueSets (elements, clone)
Renamed method emptySet to noneOf
Added additional constructors in MaskValueSetImpl

JaniruTEC committed:
Updated MaskValueSet to match EnumSet better

Added additional convenience methods to MaskValueSet (allOf, copyOf, of)
Added additional methods for copying MaskValueSets (elements, clone)
Renamed method emptySet to noneOf
Added additional constructors in MaskValueSetImpl

infeo commented on src/main/java/dev/dokan/dokan_java/masking/MaskValueSet.java:
```diff
@@ -12,10 +14,30 @@
 public interface MaskValueSet<T extends Enum<T> & MaskValueEnum> extends Set<T>, IntegerConvertible {
 
-    static <T extends Enum<T> & MaskValueEnum> MaskValueSet<T> emptySet(Class<T> clazz) {
+    static <T extends Enum<T> & MaskValueEnum> MaskValueSet<T> noneOf(Class<T> clazz) {
```
I like a coherent naming scheme, but `emptySet` is more understandable than `noneOf`.

If you add some JavaDoc, it will be okay.

JaniruTEC replied:
`EnumSet` uses `noneOf`, that's why I renamed the method; I will still change it

**EDIT:** Resolved with commit 6522e2a

infeo replied:
Ah, ok. I didn't knew about that... But now it is already changed^^

infeo commented on src/main/java/dev/dokan/dokan_java/masking/MaskValueSet.java:
```diff
@@ -12,10 +14,30 @@
+        return new MaskValueSetImpl<T>(set);
+    }
+
+    static <T extends Enum<T> & MaskValueEnum> MaskValueSet<T> copyOf(Collection<T> collection) {
```
Shouldn't this method be sufficient? It covers the other two above. 

I like your explicit approach, but we shouldn't make the classes not unnecessary fat imho. Also this introduces a lot of boilerplate code.

JaniruTEC replied:
It is sufficient. Same as with `noneOf`: EnumSet contains those overloaded methods.
I will remove them.

**EDIT:** Resolved with commit 6522e2a

infeo commented on src/main/java/dev/dokan/dokan_java/masking/MaskValueSetImpl.java:
```diff
@@ -31,6 +51,17 @@ public final void add(T... items) {
 
+    @Override
+    public EnumSet<T> elements() {
+        return this.elements.clone();
```
I don't like this method. It says "elements", but rather than just give a view on the elements it returns a copy of them. So either it should be renamed or removed.

JaniruTEC replied:
Returning a copy is a common approach to this problem tho. Editing MaskValueSet using the underlying EnumSet is not useful, but getting a copy can be.

infeo replied:
Point taken. But maybe this will be obsolete, depeding on the decision to keep this method in the implemented interface.

infeo replied:
Ah, ok. I didn't knew about that... But now it is already changed^^

infeo replied:
Point taken. But maybe this will be obsolete, depeding on the decision to keep this method in the implemented interface.

infeo commented on src/main/java/dev/dokan/dokan_java/masking/MaskValueSet.java:
```diff
@@ -51,4 +73,8 @@
 
     void add(T... items);
 
+    EnumSet<T> elements();
```
I don't understand for what use case this method is defined in the interface.

JaniruTEC replied:
Mostly internal use. Also I was under the impression that a converter method could be useful.

infeo replied:
As long as its only for internal usage and someone doesn't have a need for it, i suggest to _not_ add the method to the interface.

JaniruTEC replied:
Usually internal usage should aim to use the interfaces, too.
As I said: Converters can be useful. Giving a user the ability to use EnumSet can't be really wrong. The JDK has over 300 references to it.

JaniruTEC replied:
I assume this isn't a dealbreaker. I'm pretty sure it's better to have the method than not to have it. We can remove it later if we want. I'll merge it for now

infeo reviewed (changes requested):
I like explicitness, but on the other hand, the more code a library has, the more code must be maintained.

I'm kinda torn. It adds handy helper methods, but also adds boilerplate code. Maybe we can reduce the latter without removing the first.

JaniruTEC committed:
Renamed noneOf back to emptySet and removed unnecessary methods

Renamed method noneOf of MaskValueSet back to emptySet
Resolved: https://github.com/dokan-dev/dokan-java/pull/47#discussion_r439726314

Remove unnecessary overloaded methods and changed implementation
Resolved: https://github.com/dokan-dev/dokan-java/pull/47#discussion_r439726425
